### ISSUE TITLE
Overfitting Problem

### ISSUE THREAD

**Issue by HardeyPandya** [LXP-Never/TCNN#1]
Hello. Thanks for this model architecture code. I am just very beginner in this area and not much familiar with Deep Learnning standard procedures and all.

Actually, I have converted this code - https://github.com/haoxiangsnr/IRM-based-Speech-Enhancement-using-LSTM  into time domain and then used your model architecture code for this. I am using google Colab platform for this.

I have used TIMIT dataset which is having 8732 utterances and randomly mixed with UrbanSound8K noises at -5dB,-4dB,-3dB,-2dB,-1dB,0dB and 1dB, so I am having 8732 noisy speeches. Then I convert it into overlapping frames. And output proceeds.

But I am not sure whether it is overfitting - After 600 epochs of training validation, average PESQ score obtained is 2.13. Average PESQ between clean speech and noisy speech of UrbanSound8k + TIMIT clean speech is around 1.8 .
(On each epoch 900 noisy utterences are trained and on next epoch utterences are shuffled and it is trained on other 900 utterences)

But upon testing, I use NOIZEUS database which is unseen to the TCNN network. I am getting very low PESQ score of 1.42 after loading checkpoints. Also when i run same inference script I get different PESQ scores like 1.3, 1.4 or 1.5 something like that ! On same model checkpoint.

Any suggestions why this is the case ? It would be very helpful. 

Thanks.

**Comment by LXP-Never** [comment]
> 你好。感谢这个模型架构代码。我只是这方面的初学者，对深度学习标准程序等不太熟悉。
> 
> 实际上，我已经将此代码 - https://github.com/haoxiangsnr/IRM-based-Speech-Enhancement-using-LSTM 转换为时域，然后为此使用您的模型架构代码。我为此使用谷歌 Colab 平台。
> 
> 我使用了 TIMIT 数据集，该数据集有 8732 条话语，并随机混合了 -5dB、-4dB、-3dB、-2dB、-1dB、0dB 和 1dB 的 UrbanSound8K 噪声，因此我有 8732 条嘈杂的语音。然后我把它转换成重叠的帧。并输出收益。
> 
> 但我不确定它是否过拟合——经过 600 次训练验证后，获得的平均 PESQ 分数为 2.13。UrbanSound8k + TIMIT 干净语音的干净语音和嘈杂语音之间的平均 PESQ 约为 1.8 。 （在每个 epoch 训练 900 个嘈杂的话语，在下一个 epoch 话语被打乱，并在其他 900 个话语上训练）
> 
> 但是在测试时，我使用了 TCNN 网络看不到的 NOIZEUS 数据库。加载检查点后，我的 PESQ 分数非常低，为 1.42。此外，当我运行相同的推理脚本时，我会得到不同的 PESQ 分数，例如 1.3、1.4 或 1.5 之类的！在同一模型检查点上。
> 
> 有什么建议为什么会这样？这将非常有帮助。
> 
> 谢谢。

There are many reasons for overfitting:
1. The model complexity is too high and there are too many parameters
2. The training data set is relatively small
solution:
1. Reduce model complexity
2. More diverse data or data augmentation
3. dropout
4. Regularization
5. Stop early
6. BN

**Comment by LXP-Never** [comment]
> Hello. Thanks for this model architecture code. I am just very beginner in this area and not much familiar with Deep Learnning standard procedures and all.
> 
> Actually, I have converted this code - https://github.com/haoxiangsnr/IRM-based-Speech-Enhancement-using-LSTM into time domain and then used your model architecture code for this. I am using google Colab platform for this.
> 
> I have used TIMIT dataset which is having 8732 utterances and randomly mixed with UrbanSound8K noises at -5dB,-4dB,-3dB,-2dB,-1dB,0dB and 1dB, so I am having 8732 noisy speeches. Then I convert it into overlapping frames. And output proceeds.
> 
> But I am not sure whether it is overfitting - After 600 epochs of training validation, average PESQ score obtained is 2.13. Average PESQ between clean speech and noisy speech of UrbanSound8k + TIMIT clean speech is around 1.8 . (On each epoch 900 noisy utterences are trained and on next epoch utterences are shuffled and it is trained on other 900 utterences)
> 
> But upon testing, I use NOIZEUS database which is unseen to the TCNN network. I am getting very low PESQ score of 1.42 after loading checkpoints. Also when i run same inference script I get different PESQ scores like 1.3, 1.4 or 1.5 something like that ! On same model checkpoint.
> 
> Any suggestions why this is the case ? It would be very helpful.
> 
> Thanks.

There are many reasons for overfitting:
1. The model complexity is too high and there are too many parameters
2. The training data set is relatively small
solution:
1. Reduce model complexity
2. More diverse data or data augmentation
3. dropout
4. Regularization
5. Stop early
6. BN

**Comment by LXP-Never** [comment]
There are many reasons for overfitting:
1、The model complexity is too high and there are too many parameters
2、The training data set is relatively small
solution:
1、Reduce model complexity
2、More diverse data or data augmentation
3、dropout
4、Regularization
5、Stop early
6、BN

**Comment by HardeyPandya** [comment]
Ok thanks, but initially when training (let's say my data is too small , only 8732 utterances) , average validation PESQ score is coming less than 1 in initial epochs. Is it possible to have PESQ score less than 1.4 ? I heard it should be between 1.4 and 4.5 . So I suspect where is the problem.

I will surely try by training different dataset.

**Comment by HardeyPandya** [comment]
I will also try by reducing number of encoder decoder layers.

Also I am not sure - everytime I run the inference script loading same checkpoint and use same noisy speech I am getting different PESQ and STOI scores between predicted clean speech through trained model and clean speech. And difference is not small, it is relatively larger. Like on first run, it comes 1.49 , second run it comes 1.31 , then sometimes 1.1  and this PESQ score is even lower than the PESQ score obtained through noisy and clean speech. Why this would happen?

**Comment by LXP-Never** [comment]
PESQ is really too low, maybe you should check your training pipeline for problems? I guess

**Comment by HardeyPandya** [comment]
Yes I have checked pipeline. Converting 16kHz resampled audio speeches to overlapping frames etc, then calculating loss function and convert it back to original audio signal using Overlap and Add etc it is working properly. Maybe I am missing something it is possible. I will put colab link code soon on GitHub.

To me it seems overfitting problem to be honest and using more complex model for smaller data. But I have no knowledge of common errors and standard procedures. So that's why I am asking this question. It is very low PESQ score it is okay - but is it possible to fluctuate PESQ and STOI scores this much upon every run of inference script or is it some mistake?

Thank You for your suggestions and comments.

**Comment by HardeyPandya** [comment]
This is the code I prepared for TCNN by gathering some code segments from other sources as mentioned. 

https://github.com/HardeyPandya/Temporal-Convolutional-Neural-Network-Single-Channel-Speech-Enhancement

**Comment by LXP-Never** [comment]
👍
